### PR TITLE
Replace DatePicker component used in WeekPicker to prevent bad CSS styling (#463)

### LINKED ISSUE #463: Datetime range picker in availability page is not displaying correctly
The datetime range picker in availability page is not displaying correctly, I'm using Firefox 78 on Windows 10 2004, running 7.40.0.Final build.

图片

Also the generated all-day availability range can not be moved, and it seems that there is a problem with request's `endDateTime` property.

Related error messages (this was generated when I tried to move one time range for Amy Cole from 2020-08-14 to 2020-08-15, my current date is 2020-07-23):

<details>
<summary>Request</summary>

```
{
    "id": 2573,
    "tenantId": 1,
    "version": 0,
    "state": "UNAVAILABLE",
    "employeeId": 19,
    "startDateTime": "2020-08-15T00:00:00",
    "endDateTime": "2020-08-17T00:00:00"
}
```

</details>

<details>
<summary>Response</summary>

```
{
    "i18nKey": "ServerSideException.generic",
    "exceptionMessage": "Validation failed for classes [org.optaweb.employeerostering.domain.employee.EmployeeAvailability] during update time for groups [javax.validation.groups.Default, ]\nList of constraint violations:[\n\tConstraintViolationImpl{interpolatedMessage='只能为true', propertyPath=valid, rootBeanClass=class org.optaweb.employeerostering.domain.employee.EmployeeAvailability, messageTemplate='{javax.validation.constraints.AssertTrue.message}'}\n]",
    "messageParameters": [],
    "exceptionClass": "javax.validation.ConstraintViolationException",
    "stackTrace": [
        "org.hibernate.cfg.beanvalidation.BeanValidationEventListener.validate(BeanValidationEventListener.java:140)",
        "org.hibernate.cfg.beanvalidation.BeanValidationEventListener.onPreUpdate(BeanValidationEventListener.java:88)",
        "org.hibernate.action.internal.EntityUpdateAction.preUpdate(EntityUpdateAction.java:248)",
        "org.hibernate.action.internal.EntityUpdateAction.execute(EntityUpdateAction.java:119)",
        "org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)",
        "org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:478)",
        "org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:356)",
        "org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)",
        "org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1472)",
        "org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1458)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)",
        "java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)",
        "java.base/java.lang.reflect.Method.invoke(Method.java:566)",
        "org.springframework.orm.jpa.ExtendedEntityManagerCreator$ExtendedEntityManagerInvocationHandler.invoke(ExtendedEntityManagerCreator.java:366)",
        "com.sun.proxy.$Proxy129.flush(Unknown Source)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)",
        "java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)",
        "java.base/java.lang.reflect.Method.invoke(Method.java:566)",
        "org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:314)",
        "com.sun.proxy.$Proxy129.flush(Unknown Source)",
        "org.springframework.data.jpa.repository.support.SimpleJpaRepository.flush(SimpleJpaRepository.java:601)",
        "org.springframework.data.jpa.repository.support.SimpleJpaRepository.saveAndFlush(SimpleJpaRepository.java:570)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)",
        "java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)",
        "java.base/java.lang.reflect.Method.invoke(Method.java:566)",
        "org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:371)",
        "org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:204)",
        "org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:657)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:621)",
        "org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:605)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:80)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:366)",
        "org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:118)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:178)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:95)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)",
        "com.sun.proxy.$Proxy158.saveAndFlush(Unknown Source)",
        "org.optaweb.employeerostering.service.employee.EmployeeService.updateEmployeeAvailability(EmployeeService.java:255)",
        "org.optaweb.employeerostering.service.employee.EmployeeService$$FastClassBySpringCGLIB$$d89f3857.invoke(<generated>)",
        "org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)",
        "org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:771)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)",
        "org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)",
        "org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:366)",
        "org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:118)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)",
        "org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:691)",
        "org.optaweb.employeerostering.service.employee.EmployeeService$$EnhancerBySpringCGLIB$$bb58027f.updateEmployeeAvailability(<generated>)",
        "org.optaweb.employeerostering.service.employee.EmployeeController.updateEmployeeAvailability(EmployeeController.java:134)",
        "org.optaweb.employeerostering.service.employee.EmployeeController$$FastClassBySpringCGLIB$$156af6ba.invoke(<generated>)",
        "org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)",
        "org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:771)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)",
        "org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)",
        "org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:120)",
        "org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)",
        "org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)",
        "org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:691)",
        "org.optaweb.employeerostering.service.employee.EmployeeController$$EnhancerBySpringCGLIB$$3fbab3f9.updateEmployeeAvailability(<generated>)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)",
        "java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)",
        "java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)",
        "java.base/java.lang.reflect.Method.invoke(Method.java:566)",
        "org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)",
        "org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)",
        "org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:105)",
        "org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:879)",
        "org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793)",
        "org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)",
        "org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)",
        "org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)",
        "org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)",
        "org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:920)",
        "javax.servlet.http.HttpServlet.service(HttpServlet.java:663)",
        "org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)",
        "javax.servlet.http.HttpServlet.service(HttpServlet.java:741)",
        "org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)",
        "org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)",
        "org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)",
        "org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)",
        "org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)",
        "org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)",
        "org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)",
        "org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)",
        "org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)",
        "org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)",
        "org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)",
        "org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)",
        "org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)",
        "org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)",
        "org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)",
        "org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)",
        "org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)",
        "org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)",
        "org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)",
        "org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)",
        "org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)",
        "org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)",
        "org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)",
        "org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)",
        "org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:373)",
        "org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)",
        "org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)",
        "org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1594)",
        "org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)",
        "java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)",
        "java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)",
        "org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)",
        "java.base/java.lang.Thread.run(Thread.java:834)"
    ],
    "exceptionCause": null
}
```

</details>

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
"@wojtekmaj/react-daterange-picker" has a CSS issue causing it not to load properly at startup.
Replacing it with "react-datepicker" since we already use it in other components and it is more popular and maintained.

Fixes #463.

### CODE REVIEW

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx [hunk 1]
```diff
@@ -90,20 +118,19 @@ export default class WeekPicker extends React.Component<WeekPickerProps, WeekPic
...
+              this.goToWeekContaining(date);
+            } else {
+              this.goToCurrentWeek();
+            }
+          }}
           minDate={this.props.minDate}
           maxDate={this.props.maxDate}
-          isOpen={this.state.isOpen}
-          clearIcon={<HistoryIcon />}
+          customInput={<DateRangeInputElement />}
```
Comment (yurloc) [inline optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:133]:
```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

Check the render method of `r`.
    in DateRangeInputElement (at WeekPicker.tsx:140)
    in div (created by r)
    in div (created by InnerReference)
    in InnerReference (created by Context.Consumer)
    in Reference (created by r)
    in Manager (created by r)
    in r (created by r)
    in r (at WeekPicker.tsx:126)
    in div (at WeekPicker.tsx:100)
    in WeekPicker (at AvailabilityRosterPage.tsx:393)
    in span (at AvailabilityRosterPage.tsx:370)
    in AvailabilityRosterPage (created by Context.Consumer)
    in withRouter(AvailabilityRosterPage) (created by ConnectFunction)
    in ConnectFunction (created by withI18nextTranslation(Connect(withRouter(AvailabilityRosterPage))))
    in withI18nextTranslation(Connect(withRouter(AvailabilityRosterPage))) (created by Context.Consumer)
    in Route (at App.tsx:86)
    in Switch (at App.tsx:55)
    in section (created by PageSection)
    in PageSection (at App.tsx:45)
    in main (created by Page)
    in div (created by Page)
    in Page (at App.tsx:39)
    in App (at src/index.tsx:59)
    in Suspense (at src/index.tsx:58)
    in Router (created by BrowserRouter)
    in BrowserRouter (at src/index.tsx:57)
    in Provider (at src/index.tsx:56)
    in I18nextProvider (at src/index.tsx:55)
```
If I understand it correctly, `DateRangeInputElement` is a function component but the render method of `DatePicker` tries to pass a ref to it, which is kinda not allowed. This looks fishy. Can you try fixing this?

A similar situation is described and solved in https://stackoverflow.com/questions/58574298/warning-with-function-components-cannot-be-given-refs-even-though-using-forwar.

Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:133]: Their custom input example does not use useForwardRef: https://reactdatepicker.com/. However, I can confirm using forwardRef fixes the error.

Comment (yurloc) [reply optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:143]: Perfect, thanks!

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx [hunk 2]
```diff
@@ -40,28 +35,71 @@ function getLastDayInWeek(dateInWeek: Date): Date {
...
   }
 
   render() {
-    const locale = moment.locale();
+    type DatePickerCustomInputProps = { value?: Date; onClick?: () => void };
+
+    // (the stateless functional component is defined in a class component, where
+    //  this = the class this, so it okay to use "this" here)
+    /* eslint-disable react/no-this-in-sfc */
+    const DateRangeInputElement = React.forwardRef<HTMLDivElement, DatePickerCustomInputProps>(
```
Comment (yurloc) [inline optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:60]: Could you please move this component to a dedicated file? The render method is already pretty long so moving this sub-component outside will make the WeekPicker easier to understand. Also, the move will help to get rid of the references to `this` and replace them by props. A function component output should be fully determined by the input props (and perhaps its state).

More about this at https://reactjs.org/docs/components-and-props.html.

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx [hunk 3]
```diff
@@ -40,28 +35,71 @@ function getLastDayInWeek(dateInWeek: Date): Date {
...
+          className="pf-c-form-control"
+          style={{
+            display: 'grid',
+            height: 'auto',
+            gridTemplateColumns: '1fr auto auto',
+            alignItems: 'center',
+          }}
+        >
+          <Text
+            onClick={dateRangeInputProps ? dateRangeInputProps.onClick : undefined}
```
Comment (yurloc) [inline optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:73]: Do you know how the `onClick` works? I wonder if it is possible to make it work like a toggle. First click should open the date picker, second click should close it. If I click the text accidentally or change my mind and decide not to select a date it is now impossible to hide the data picker with another click at the same place. For me it's a bad user experience.

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx [hunk 4]
```diff
@@ -40,28 +35,71 @@ function getLastDayInWeek(dateInWeek: Date): Date {
...
   goToCurrentWeek() {
     this.goToWeekContaining(new Date());
   }
 
   goToWeekContaining(date: Date) {
     if (this.props.minDate && getFirstDayInWeek(date) < getFirstDayInWeek(this.props.minDate)) {
       this.goToWeekContaining(this.props.minDate);
-    } else if (this.props.maxDate && getLastDayInWeek(date) > getLastDayInWeek(this.props.maxDate)) {
+      return;
+    } if (this.props.maxDate && getLastDayInWeek(date) > getLastDayInWeek(this.props.maxDate)) {
```
Comment (yurloc) [inline optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:47]: `if` should start on a new line.

Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:47]: I wonder why this isn't an eslint violation; I didn't get any errors or warnings when running `npm run lint`

**Comment by yurloc** [comment]
@Christopher-Chianelli please also remove `react-date-picker`. It's not used directly by our code so it shouldn't declared in `package.json`. Perhaps it was used in the past and the dependency wasn't removed. On the other hand, it's a dependency of `@wojtekmaj/react-daterange-picker` so it makes sense to remove it in this PR.

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx (hunk 2 above)
Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx:60]: I disagree with moving it to another file; the custom input component is customized for use with the week picker component and wouldn't make sense using it in another component/page. I can extract the component though (won't be in the class render method, but be in the same file).

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx (hunk 3 above)
Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx]: I managed to enable toggle behaviour by getting the DatePicker ref and calling `setOpen`, keeping track of the `isOpen` state in the custom input component.

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx (hunk 2 above)
Comment (yurloc) [reply optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx]: That also works.

Generally, I don't think that reusing a component more than once is a necessary requirement for moving it to a separate file. One component per file helps to keep the file size limited making it easier to read and focus on a specific problem the component solves.

File: optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx (hunk 3 above)
Comment (yurloc) [reply optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx]: Great, thanks!

**Pull request merged**